### DISCUSSION TITLE
how to check if a provider/stateNotifier exists in the ProviderScope.

### DISCUSSION THREAD

**Discussion by Hasankanso** [rrousselGit/riverpod#1426] [Q&A] [answered]
Hello there,

```
final counterProvider = StateNotifierProvider.family<Counter, int, int>((ref, id) {
  return Counter(id);
});

class Counter extends StateNotifier<int> {
  final int id;
  Counter(this.id) : super(0);
  void increment() => state++;
}

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {

  @override
  Widget build(BuildContext context, WidgetRef ref) {

    final count = ref.watch(counterProvider(4)); // counter with id 4 is created.
    
    return ...
      home: Scaffold( floatingActionButton: 
           ... 
            onPressed: () {
             int id = 2;
             //need the implementation of this method
              if(counterWithIdExists(id)){ 
                   //normally, it will not be executed, cause counter with id 2 is not yet created.
                       ref.read(counterProvider(id)).increment(); 
            },
          ),)  ),
    );
  }
}
```

how do I check the existence of a provider, since read and watch will create it if it wasn't there? (the code above is an example)

thank you for your time.

**Comment by devnta** [comment]
Is there an answer to this question?

**Reply by Hasankanso** [reply]
> Is there an answer to this question?

No, I changed on my code's structure, so I don't need it anymore, it could be a bad design that you need to check the existence of the stateNotifier (just a gut feeling).

Anyway, as a workaround, you can have a bool variable in the stateNotifier/state class, initially false. Change the variable to true when you know that you'll need that stateNotifier. Then when you want to make sure if a stateNotifier is used before (exists), call ref.read and check the variable if it is true, if not, then dispose the newly created stateNotifier (I never needed to dispose a stateNotifier, so I don't know how to do it).

this is what I could think of. There could be a built in solution, don't know...

**Comment by Jayshanx** [comment]
try this

```
const String _countProviderName = 'kCounterProvider';

final counterProvider = StateNotifierProvider.family<Counter, int, int>(
      (ref, id) {
    return Counter(id);
  },
  name: _countProviderName, //<=== here add name
);


//here check if Provider exist
bool counterWithIdExists(BuildContext context, int id) {
  return ProviderScope.containerOf(context).getAllProviderElements().any((element) {
    return element.provider.name == _countProviderName && element.provider.argument == id;
  });
}
```

**Comment by rrousselGit** [comment, marked as answer]
With 2.1.0, you will now be able to do:

```dart
ref.exists(provider);
container.exists(provider);
widgetRef.exists(provider);
```

**Reply by gajendrapandeya** [reply]
What if i am using earlier version of riverpod?

**Reply by rrousselGit** [reply]
Upgrade riverpod.

**Reply by gajendrapandeya** [reply]
Anyother solutions 
As i am working with older version of dart and fluttter
